### PR TITLE
AG-10565/10567: Add an enterprise SyncModule

### PR DESCRIPTION
**Pull request by iMoses**
Missing validations but otherwise working

### CODE REVIEW

File: packages/ag-charts-community/src/chart/interaction/syncManager.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import type { Chart } from '../chart';
+import { BaseManager } from './baseManager';
+
+type GroupId = string | symbol;
+
+export class SyncManager extends BaseManager {
+    static chartsGroups = new Map<GroupId, Set<Chart>>();
+    static DEFAULT_GROUP = Symbol('sync-group-default');
```
Comment (alantreadway) [inline packages/ag-charts-community/src/chart/interaction/syncManager.ts:8]: Minor nit: Any reason to not just make this a `const` outside the class definition?

File: packages/ag-charts-community/src/module-support.ts [hunk 2]
```diff
@@ -38,6 +38,7 @@ export * from './chart/interaction/interactionManager';
 export * from './chart/interaction/gestureDetector';
 export * from './chart/interaction/tooltipManager';
 export * from './chart/interaction/zoomManager';
+export { type Chart } from './chart/chart';
```
Comment (alantreadway) [inline packages/ag-charts-community/src/module-support.ts]: I don't like exporting `Chart`, as it's interface contract is a bit too wild and unrestricted - could we introduced a narrower interface contract for sync-related methods which `Chart` implements?

File: packages/ag-charts-community/src/options/chart/chartOptions.ts [hunk 3]
```diff
@@ -147,6 +154,8 @@ export interface AgBaseThemeableChartOptions<TDatum = any> {
     zoom?: AgZoomOptions;
     /** Configuration for the Navigator. */
     navigator?: AgNavigatorOptions;
+    /** Configuration for synchronizing multiple charts. */
+    sync?: AgChartSyncOptions;
```
Comment (alantreadway) [inline packages/ag-charts-community/src/options/chart/chartOptions.ts]: Could we hold off adding this to the options contract until we have done a few iterations with David and agreed the options contract?

File: packages/ag-charts-enterprise/src/features/sync/chartSync.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        }
+    }
+
+    private enabledZoomSync() {
+        const { syncManager, zoomManager } = this.moduleContext;
+        this.disableZoomSync = zoomManager.addListener('zoom-change', (e) => {
+            if ('stopPropagation' in e && e.stopPropagation) return;
+            for (const chart of syncManager.getSiblings(this.groupId)) {
+                chart.zoomManager.updateZoom('zoom', zoomManager.getZoom());
+                (chart.zoomManager as any).listeners.dispatch('zoom-change', { ...e, stopPropagation: true });
```
Comment (alantreadway) [inline packages/ag-charts-enterprise/src/features/sync/chartSync.ts]: Shouldn't `updateZoom()` percolate the change itself?

File: packages/ag-charts-enterprise/src/features/sync/chartSync.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        const { syncManager } = this.moduleContext;
+        const chart = syncManager.getChart();
+        const syncSeries = syncManager.getGroup(this.groupId).flatMap((chart) => chart.series);
+
+        for (const axis of chart.axes) {
+            if (this.axes !== 'xy' && this.axes !== axis.direction) continue;
+            axis.boundSeries = syncSeries.filter((series) => {
+                const seriesKeys = series.getKeys(axis.direction);
+                return axis.keys.length ? axis.keys.some((key) => seriesKeys?.includes(key)) : true;
+            });
```
Comment (alantreadway) [inline packages/ag-charts-enterprise/src/features/sync/chartSync.ts:73]: I have some concerns about the scalability of this, every chart axis instance has references to every other charts series?

File: packages/ag-charts-community/src/chart/interaction/syncManager.ts (hunk 1 above)
Comment (iMoses) [reply packages/ag-charts-community/src/chart/interaction/syncManager.ts:8]: Matter of opinion, nothing more. I like that consts of specific classes are attached to them, creates sort of a namespace for the relevant consts. Downside is that you cannot import them directly.
You want me to change it?

File: packages/ag-charts-community/src/options/chart/chartOptions.ts (hunk 3 above)
Comment (iMoses) [reply packages/ag-charts-community/src/options/chart/chartOptions.ts]: I'll comment it out

File: packages/ag-charts-community/src/module-support.ts (hunk 2 above)
Comment (iMoses) [reply packages/ag-charts-community/src/module-support.ts]: I think I ended up not using it anyhow. I'll remove it.

File: packages/ag-charts-enterprise/src/features/sync/chartSync.ts (hunk 5 above)
Comment (iMoses) [reply packages/ag-charts-enterprise/src/features/sync/chartSync.ts:73]: It is currently only used for calculating the domain, and there's nothing more to it.
I agree it's not ideal but I believe the solution is a refactor of the axis and chart implementation.

File: packages/ag-charts-enterprise/src/features/sync/chartSync.ts (hunk 4 above)
Comment (iMoses) [reply packages/ag-charts-enterprise/src/features/sync/chartSync.ts]: you're right. good catch 👍

File: packages/ag-charts-community/src/chart/interaction/syncManager.ts (hunk 1 above)
Comment (alantreadway) [reply packages/ag-charts-community/src/chart/interaction/syncManager.ts:8]: :+1:

File: packages/ag-charts-enterprise/src/features/sync/chartSync.ts (hunk 5 above)
Comment (alantreadway) [reply packages/ag-charts-enterprise/src/features/sync/chartSync.ts:73]: Discussed offline, we should follow up with a refactor to try and decouple axis & scales, so we could avoid O(n^3) complexity in the domain calculation.

**Review by alantreadway — APPROVED** [review]
LGTM

**Pull request merged**